mm2src/coins/eth.rs

```diff
@@ -4251,22 +4309,25 @@ impl EthCoin {
     /// because [`CallRequest::from`] is set to [`EthCoinImpl::my_address`].
     fn estimate_gas_for_contract_call(&self, contract_addr: Address, call_data: Bytes) -> Web3RpcFut<U256> {
         let coin = self.clone();
-        Box::new(coin.get_gas_price().and_then(move |gas_price| {
-            let eth_value = U256::zero();
-            let estimate_gas_req = CallRequest {
-                value: Some(eth_value),
-                data: Some(call_data),
-                from: Some(coin.my_address),
-                to: Some(contract_addr),
-                gas: None,
-                // gas price must be supplied because some smart contracts base their
-                // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
-                gas_price: Some(gas_price),
-                ..CallRequest::default()
-            };
-            coin.estimate_gas_wrapper(estimate_gas_req)
-                .map_to_mm_fut(Web3RpcError::from)
-        }))
+        let fee_policy_for_estimate = get_swap_fee_policy_for_estimate(self.get_swap_transaction_fee_policy());
+        Box::new(
+            coin.get_swap_pay_for_gas_option(fee_policy_for_estimate)
+                .and_then(move |pay_for_gas_option| {
+                    let eth_value = U256::zero();
+                    let estimate_gas_req = CallRequest {
+                        value: Some(eth_value),
+                        data: Some(call_data),
+                        from: Some(coin.my_address),
+                        to: Some(contract_addr),
+                        ..CallRequest::default()
+                    };
+                    // gas price must be supplied because some smart contracts base their
+                    // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
+                    let estimate_gas_req = call_request_with_pay_for_gas_option(estimate_gas_req, pay_for_gas_option);
+                    coin.estimate_gas_wrapper(estimate_gas_req)
+                        .map_to_mm_fut(Web3RpcError::from)
+                }),
+        )
     }
```

Comment:
Please make this function async and remove `and_then` since it can be removed after making the function async.